### PR TITLE
[ROCm][DSv4.1] Keep the Engram tables in host memory on ROCm

### PR DESCRIPTION
**Pull request by JohnQinAMD**
## Purpose

DeepSeek-V4.1-Flash's Engram n-gram tables are the largest single consumer of device memory — 47.2 GiB per rank at TP=4, about a third of everything the weights hold. CUDA moves them to pinned host memory and reads them over UVA (#56512); on ROCm two `is_cuda()` gates turn that off, so `engram_config` stays `None`, `amd/model.py` falls through to the base `Engram`, and the tables stay on the accelerator.

Nothing in the offload path is platform-specific — `get_accelerator_view_from_cpu_tensor()` already dispatches on `is_cuda_alike()`, and `hipHostRegister` backs pinned host memory — so widen the two gates and import `Engram` from `..nvidia.engram`, as `amd/model.py` already does for `DeepseekV41ModelState`. No new AMD code. The `cpu_offload` tests in `tests/kernels/test_engram.py` carried the same gate and are widened too. `DPSharedEngramStorage` and the DP head sharding arrive with the import but stay inert at `data_parallel_size=1`.

Duplicate-work check (2026-09-18): #56512 (merged) built the NVIDIA offload and left these gates behind; #56926, #57195 and #56757 continue that path and none touches `amd/` or the gates — #56757's `disk_offload_dir` requires `cpu_offload=True`, so it is blocked on ROCm by the same gate. #56400 has an AMD/ROCm section with no engram item. Prior art: SemiAnalysisAI/InferenceX#3217 by @functionstackx reached the same design as a local patch for an MI355X TP2 recipe, and recorded that hardware evidence was the outstanding gate before upstreaming. `vllm/config/engram.py` is also edited by those three, so this needs a rebase onto whichever merges first.

## Test plan

```bash
python -m pytest tests/kernels/test_engram.py -q
python -m ruff check vllm/config/engram.py vllm/config/vllm.py \
  vllm/models/deepseek_v41/amd/model.py tests/kernels/test_engram.py
python -m ruff format --check vllm/config/engram.py vllm/config/vllm.py \
  vllm/models/deepseek_v41/amd/model.py tests/kernels/test_engram.py
git diff --check

# MI355X (gfx950), ROCm 7.2.3, TP=4, one session per arm, warmup then 3 reps
vllm serve $MODEL --tensor-parallel-size 4 --language-model-only \
  --tokenizer-mode deepseek_v41 --trust-remote-code --moe-backend aiter \
  --gpu-memory-utilization 0.9 --max-model-len 33792 --max-num-seqs 128 \
  --max-cudagraph-capture-size 1024 --max-num-batched-tokens 16384
vllm bench serve --backend vllm --endpoint /v1/completions --model $MODEL \
  --dataset-name random --random-input-len 4096 --random-output-len 512 \
  --random-range-ratio 0 --num-prompts 32 --max-concurrency 8 --seed 1234 \
  --ignore-eos
python tests/evals/gsm8k/gsm8k_eval.py --port 8000 --num-questions 1319 \
  --num-shots 5 --max-concurrency 64

# DP shared storage, TP=2 / DP=2
vllm serve $MODEL --tensor-parallel-size 2 --data-parallel-size 2 \
  --engram-config '{"cpu_offload":true,"dp_shared_memory":true}' ...
```

## Test result

- `tests/kernels/test_engram.py` on gfx950: the two `cpu_offload` tests are the reason for the test change in this PR. Their gates were `is_cuda()`, so on ROCm the whole file was 9 passed / 75 skipped and the offload path had no coverage at all. Widened to `is_cuda_alike()`, `test_engram_lookup_matches_torch` (2 cpu_offload x 2 background x 3 token counts) and `test_engram_head_shards_reconstruct_checkpoint` (2 cpu_offload x 4 TP sizes) give **20 passed**, including the bit-exact comparison against the torch dequant reference reading from pinned host memory.
- The alias is coherent on gfx950, which is what the stricter `is_uva_available()` proposed in #57404 probes for: host writes made after the accelerator view is taken are visible to the device, and device writes are visible to the host, for uint8 and float8_e4m3fn at 4 KiB and 1 MiB.
- ruff check, ruff format --check and `git diff --check` pass on all four files. mypy and pre-commit were not run — neither is installed on the test host.
- TP=4 serving: the offload log line appears once per rank per layer (8 lines, 23.60 GiB each), and `Resolved Engram configuration` appears at all, which it does not on base.
- TP=2 / DP=2 with `dp_shared_memory`: server comes up and serves, four shared allocations of 47.21 GiB logged as "shared across DP replicas", 190 GiB resident in `/dev/shm`, no errors. That run is also the only one that reaches `verify_model_config()`, the second gate, since only an explicitly supplied `--engram-config` gets there. Functional check only; no DP performance claim is made.

## Model evaluation

`tests/evals/gsm8k/gsm8k_eval.py`, all 1319 problems, five-shot, concurrency 64, MI355X TP=4:

| | accuracy | invalid rate |
| --- | ---: | ---: |
| base | 0.8999 | 0.15% |
| this PR | 0.9030 | 0.08% |

+0.31 pp against a 1.16 pp standard error for the difference of two proportions, z = 0.27. The acceptance band was fixed before the run. A second offload build measured 0.9052 (z = 0.46); see the note under Performance.

## Performance

The prefetch is what keeps this free. In the shape Engram actually issues — 3 n-gram orders x 6 local hash heads = 18 rows of 256 B per token, in 2 of 40 layers — a UVA gather matches HBM at decode batch sizes and falls behind only at prefill batch sizes:

| tokens/step | rows | UVA view | HBM | delta |
| ---: | ---: | ---: | ---: | ---: |
| 1 | 18 | 9.3 us | 10.7 us | -1.4 us |
| 32 | 576 | 9.9 us | 9.8 us | +0.1 us |
| 256 | 4608 | 25.6 us | 9.9 us | +15.7 us |
| 1024 | 18432 | 89.0 us | 10.3 us | +78.7 us |

The model-owned prefetch stream, with `background=True` leaving half the SMs to the main stream, covers the last two rows.

**KV cache capacity**, TP=4, identical across three independent starts:

| | base | this PR | delta |
| --- | ---: | ---: | ---: |
| Available KV cache memory | 107.8 GiB | 154.38 GiB | **+43.2%** |
| GPU KV cache size | 4,325,113 tokens | 6,194,599 tokens | **+43.2%** |
| Max concurrency @ 33,792 ctx | 127.99x | 183.32x | **+43.2%** |

Predicted from the startup memory breakdown beforehand as +43.8%; the measurement met it within 0.4%.

At TP=2 / DP=2 each rank carries a wider engram shard (47.21 GiB per layer instead of 23.60), so the same change is correspondingly larger. Both arms are the same topology and flags; the base arm simply cannot be given `--engram-config`, because that is the gate this PR widens:

| | base | this PR | delta |
| --- | ---: | ---: | ---: |
| Available KV cache memory | 57.65 GiB | 151.45 GiB | **+162.7%** |
| GPU KV cache size | 2,313,707 tokens | 6,078,307 tokens | **+162.7%** |

The startup breakdown pins down what moved: per rank, "consumed memory (weights + non-torch)" is 182.05 GiB on base and 87.62 GiB with the offload, a difference of 94.43 GiB against the 94.42 GiB the two engram shards occupy at this TP.

**Serving throughput**, TP=4, random 4096 in / 512 out, 32 prompts, concurrency 8, three repetitions each (mean +/- sd). "paired" is the offload arm measured back to back with base in one session; "shipped" is this PR's exact diff, measured separately:

| | base | offload, paired | offload, shipped |
| --- | ---: | ---: | ---: |
| mean TPOT | 11.35 +/- 0.02 ms | 11.35 +/- 0.02 ms | 11.35 +/- 0.00 ms |
| median TPOT | 11.22 +/- 0.01 ms | 11.24 +/- 0.02 ms | 11.24 +/- 0.00 ms |
| p99 TPOT | 12.54 +/- 0.01 ms | 12.54 +/- 0.02 ms | 12.56 +/- 0.01 ms |
| mean TTFT | 796.88 +/- 2.43 ms | 794.08 +/- 0.35 ms | 799.97 +/- 1.61 ms |
| p99 TTFT | 1061.31 +/- 1.69 ms | 1054.38 +/- 2.59 ms | 1063.68 +/- 4.60 ms |
| output throughput | 620.64 +/- 0.97 tok/s | 621.10 +/- 1.06 tok/s | 620.61 +/- 0.11 tok/s |

Paired against base: +0.46 tok/s on a pooled standard error of 0.83, t = 0.55. Shipped against the same base: -0.03 tok/s, -0.005%. Flat either way, p99 included. There are two offload columns because the paired arm reached the offload path through the local copy of `nvidia/engram.py` described above, which is not the diff here; the shipped diff reproduces it exactly, down to the same KV cache figures and offload log lines.

Host cost at TP=4: 47.2 GiB x 4 ranks = 188.8 GiB of pinned host memory, against 2954 GiB of RAM on the test node. Containers need `--ulimit memlock=-1`.

AI assistance was used for this change. Every changed line was reviewed by the submitter, and the measurements above were produced by the submitter on the hardware named.

### CODE REVIEW

**Comment by functionstackx** [comment]
hi @JohnQinAMD thanks for the PR, do have an sweep of the the perf with thsi patch or do u want me to sweep it?

with my ai vibecode i got this https://github.com/SemiAnalysisAI/InferenceX/pull/3217

image

**Comment by JohnQinAMD** [comment]
> hi @JohnQinAMD thanks for the PR, do have an sweep of the the perf with thsi patch or do u want me to sweep it?
> 
> with my ai vibecode i got this [SemiAnalysisAI/InferenceX#3217](https://github.com/SemiAnalysisAI/InferenceX/pull/3217)
> 
> image

@functionstackx  It will be great if you can help sweep the perf. Thanks!

**Comment by functionstackx** [comment]
@JohnQinAMD on it

**Review by Fangzhou-Ai — APPROVED** [review]

**Comment by functionstackx** [comment]
@JohnQinAMD i dont think evals pass, let me see if this an flake

**Comment by Fangzhou-Ai** [comment]
> @JohnQinAMD i dont think evals pass, let me see if this an flake
> 
> 

@functionstackx it could be a kernel side issue that has been fixed in AITER, we are upgrading the AITER version to solve it. A hot patch was raised before but has negative performance impact so it's deprecated. The offloading PR itself can proceed as it's orthogonal and important to TP2 performance.

**Comment by functionstackx** [comment]
> > @JohnQinAMD i dont think evals pass, let me see if this an flake
> 
> @functionstackx it could be a kernel side issue that has been fixed in AITER, we are upgrading the AITER version to solve it. A hot patch was raised before but has negative performance impact so it's deprecated. The offloading PR itself can proceed as it's orthogonal and important to TP2 performance.

let me try retriggering to see if it is an flake cuz my OG patch had passing evals but this one doesnt have passing evals yet

**Comment by Fangzhou-Ai** [comment]
> > > @JohnQinAMD i dont think evals pass, let me see if this an flake
> 
> > 
> 
> > @functionstackx it could be a kernel side issue that has been fixed in AITER, we are upgrading the AITER version to solve it. A hot patch was raised before but has negative performance impact so it's deprecated. The offloading PR itself can proceed as it's orthogonal and important to TP2 performance.
> 
> 
> 
> let me try retriggering to see if it is an flake cuz my OG patch had passing evals but this one doesnt have passing evals yet

Thanks for your cross validation!

**Review by Fangzhou-Ai — COMMENTED** [review]
Waiting for more test results, PR pending

**Comment by JohnQinAMD** [comment]
@functionstackx  @Fangzhou-Ai   On the accuracy question: I ran GSM8K on both arms with two independent harnesses, and the offload change does not move the score.

Image `vllm/vllm-openai-rocm:nightly-0bfc7a15d095fe83ecc82b50561a93c177fece2d` (this branch's merge base), 8x MI355X, TP=4, `--moe-backend aiter`. `base` mounts nothing; `offload` bind-mounts only the three changed non-test files.

<details>
<summary>How it was launched</summary>

```bash
IMG=vllm/vllm-openai-rocm:nightly-0bfc7a15d095fe83ecc82b50561a93c177fece2d
VP=/usr/local/lib/python3.12/dist-packages/vllm

docker run --rm --network host --ipc host --shm-size 32g   --ulimit memlock=-1 --ulimit stack=67108864   --device /dev/kfd --device /dev/dri --group-add video   --cap-add SYS_PTRACE --security-opt seccomp=unconfined   -e ROCR_VISIBLE_DEVICES=0,1,2,3   -e NCCL_SOCKET_IFNAME=eno0 -e NCCL_IB_DISABLE=1   -e VLLM_ROCM_USE_AITER=1 -e VLLM_ROCM_USE_AITER_MOE=1   -e VLLM_USE_BREAKABLE_CUDAGRAPH=1 -e OMP_NUM_THREADS=1   -e MODEL=deepseek-ai/DeepSeek-V4.1-Flash -e MODEL_PREFIX=dsv41flash   -e TP=4 -e CONC=8 -e KV_OFFLOADING=none -e TOTAL_CPU_DRAM_GB=3023   -e EVAL_ONLY=true -e IS_AGENTIC=true -e PORT=8000   -e PP_SIZE=1 -e PCP_SIZE=1 -e REQUIRE_POWER=false -e IS_MULTINODE=false   -e RESULT_DIR=/out/results -e RESULT_FILENAME=result.json   -e INFMAX_CONTAINER_WORKSPACE=/work/InferenceX   -e HF_HOME=/hf-cache -e HF_HUB_CACHE=/hf-cache/hub   -v $INFERENCEX:/work/InferenceX -v $OUT:/out -v $HF_CACHE:/hf-cache   -v $PRARM/rocm_aiter_moe.patched.py:$VP/model_executor/layers/fused_moe/experts/rocm_aiter_moe.py:ro   `# the three lines below are the only difference between the arms; base has none`   -v $WT/vllm/config/engram.py:$VP/config/engram.py:ro   -v $WT/vllm/config/vllm.py:$VP/config/vllm.py:ro   -v $WT/vllm/models/deepseek_v41/amd/model.py:$VP/models/deepseek_v41/amd/model.py:ro   --entrypoint bash "$IMG" -c '
    cd /work/InferenceX
    set -a; source benchmarks/runtime_settings.sh; set +a
    export ENABLE_AGENTX_POWER=0
    bash benchmarks/single_node/agentic/dsv41flash_fp4_mi355x_vllm_mtp.sh
  '
```

The recipe then builds the server command itself; both arms came out byte-identical
(from the `vllm_command.txt` it writes):

```
vllm serve deepseek-ai/DeepSeek-V4.1-Flash --host 0.0.0.0 --port 8000   --tensor-parallel-size 4 --language-model-only --tokenizer-mode deepseek_v41   --tool-call-parser deepseek_v41 --enable-auto-tool-choice   --reasoning-parser deepseek_v41 --moe-backend aiter   --gpu-memory-utilization 0.9 --max-model-len 1048576 --max-num-seqs 128   --max-cudagraph-capture-size 1024 --max-num-batched-tokens 16384   --disable-uvicorn-access-log   --speculative-config '{"method":"dspark","num_speculative_tokens":5,
    "draft_sample_method":"probabilistic","rejection_sample_method":"block",
    "enable_adaptive_verification":false}'
```

`EVAL_ONLY=true` is what puts `rejection_sample_method` on `block` rather than the
`synthetic` / AL 3.51 the throughput runs use. `NCCL_SOCKET_IFNAME` and
`--ulimit memlock=-1` are needed on this cluster, not by this patch.

Both arms also carry a backport of #57036, which is still open and not on main yet (the equivalent fix is in https://github.com/ROCm/aiter/pull/5295). Without
it the AITER fused-expert path indexes expert weights with the -1 cudagraph padding
sentinel and every rank faults during startup profiling, so the server will not comes up :

```python
# vllm/model_executor/layers/fused_moe/experts/rocm_aiter_moe.py
is_padding = topk_ids < 0
topk_ids = torch.where(is_padding, torch.zeros_like(topk_ids), topk_ids)
topk_weights = torch.where(is_padding, torch.zeros_like(topk_weights), topk_weights)
```

</details>

InferenceX recipe with `EVAL_ONLY=true` (chat template, `#### [number]` instruction, two regex extractors, 5-shot greedy, all 1319 problems):

| arm | flexible-extract | strict-match |
| --- | ---: | ---: |
| base | 0.9719 +/- 0.0045 | 0.9727 +/- 0.0045 |
| offload | 0.9742 +/- 0.0044 | 0.9750 +/- 0.0043 |

Both clear `infx/evals/thresholds.yaml` `default.gsm8k = 0.90` by about 7 pp, and the two arms differ by +0.23 pp.

vLLM's own harness as a cross-check, raw `/v1/completions`, no chat template (`tests/evals/gsm8k/gsm8k_eval.py --num-questions 1319 --num-shots 5 --max-concurrency 64`):

| arm | accuracy | invalid rate |
| --- | ---: | ---: |
| base | 0.8999 | 0.15% |
| offload | 0.9030 | 0.08% |

**Review by Fangzhou-Ai — APPROVED** [review]

**Comment by Fangzhou-Ai** [comment]
PR good to go

**Comment by functionstackx** [comment]
passed on my end too https://github.com/SemiAnalysisAI/InferenceX/actions/runs/35310325966/job/105512397812?pr=3217

thank you @JohnQinAMD @Fangzhou-Ai  for turning on PoC patch into an PR that is hopefully mergable

**Review by functionstackx — APPROVED** [review]

**Comment by functionstackx** [comment]
image

**Comment by functionstackx** [comment]
@hongxiayang can we get an merge on this? 10-20% better perf with this PR

**Review by shen-shanshan — APPROVED** [review]
Reviewed by @Fangzhou-Ai
